C++-ification of src/lib/util/opresolv.[cpp|h]

npwoods opened the pull request:
Moved to util:: namespace
Changed option_resolution to be a class
C++ comments
Fixed client sites

cuavas commented on src/lib/util/opresolv.cpp:
```diff
@@ -410,51 +388,56 @@ static const struct option_resolution_entry *option_resolution_lookup_entry(opti
+// -------------------------------------------------
 
-int option_resolution_countoptions(const option_guide *guide, const char *specification)
+int option_resolution::count_options(const option_guide *guide, const char *specification)
```
this should probably return an unsigned integer or a `std::size_t` rather than a signed integer since a negative result would be meaningless

cuavas commented:
There are a number of functions that take a pointer parameter but take for granted that it will never be null.  If the parameter can really never be null, it should be a reference.  If it can be null, then the functions should check for `nullptr` and deal with it appropriately.

If it's an implementation of an interface but for some reason something up the chain should always be setting it for this particular implementation of the interface, then the interface probably needs re-thinking.  At the very least it should assert that the pointer is non-null and have a comment explaining why.

npwoods commented:
I changed the int to size_t

I've added assertions to where we were assuming that an option_resolution was provided, but held off on deeper changes.  This is because I expect to make more intense changes in the future; it isn't clear to me that the format index should be a separate parameter, for example.
